### PR TITLE
feat: add getPublicKeyFromInput

### LINKED ISSUE #11: Add `getPublicKeyFromInput()` to `Transaction` class
Add the following function to `Transaction` class
```js
getPublicKeyFromInput(index: number): Buffer | null {}
```

This function should extract the public key from the `script` or `witness` property of the input whichever is relevant based on the input type.
The following output script types contain public key in `script`:

- P2PKH

The following output script types contain public key in `witness`:

- P2WPKH
- P2SH-P2WPKH

For all other script types, the function should return `null`

### PR DESCRIPTION
**Pull request by emjshrx**
fixes #11

### CODE REVIEW

File: packages/core/test/fixtures/transaction.ts [hunk 1]
```diff
@@ -597,6 +647,7 @@ export const validTransactions = [
                     script: '76a914851a33a5ef0d4279bd5854949174e2c65b1d450088ac',
                 },
             ],
+            inPubkeys: [null],
```
Comment (theanmolsharma) [inline packages/core/test/fixtures/transaction.ts:650]: This is not right. It should not be null.
Pub key is `038de63cf582d058a399a176825c045672d5ff8ea25b64d28d4375dcdb14c02b2b`

```suggestion
            inPubkeys: ['038de63cf582d058a399a176825c045672d5ff8ea25b64d28d4375dcdb14c02b2b'],
```

File: packages/core/src/transaction.ts [hunk 2]
```diff
@@ -126,4 +126,37 @@ export class Transaction {
...
+        // p2wpkh
+        if (scriptVector.length == 0 && input.witness.length === 2) {
+            return input.witness[1];
+        }
+
+        // p2sh-p2wpkh
+        if (
+            scriptVector.length == 3 &&
+            input.script[0] === 0x00 &&
+            input.witness.length === 2
```
Comment (theanmolsharma) [inline packages/core/src/transaction.ts:156]: I think this should fix it

```suggestion
            input.script.length === 23 &&
            scriptVector.length == 1 &&
            scriptVector[0][0] === 0x00 &&
            scriptVector[0][1] === 0x14 &&
            input.witness.length === 2 &&
            input.witness[1].length === 33
```

File: packages/core/src/transaction.ts [hunk 3]
```diff
@@ -126,4 +126,37 @@ export class Transaction {
...
+
+        // p2pkh
+        if (scriptVector.length === 2) {
+            return scriptVector[1];
+        }
+
+        // p2wpkh
+        if (scriptVector.length == 0 && input.witness.length === 2) {
+            return input.witness[1];
+        }
```
Comment (theanmolsharma) [inline packages/core/src/transaction.ts:150]: Please add a check to ensure the second element is a pub key. Maybe compare length?

File: packages/core/src/transaction.ts [hunk 4]
```diff
@@ -126,4 +126,37 @@ export class Transaction {
...
+        const buffer = input.script;
+        for (let offset = 0; offset < buffer.byteLength; ) {
+            const sigItemLen = readVarInt(buffer, offset);
+            offset += encodingLength(sigItemLen);
+            scriptVector.push(buffer.subarray(offset, offset + sigItemLen));
+            offset += sigItemLen;
+        }
+
+        // p2pkh
+        if (scriptVector.length === 2) {
```
Comment (theanmolsharma) [inline packages/core/src/transaction.ts:143]: same comment as below. Make sure it's a pub key.

File: packages/core/test/fixtures/transaction.ts
Comment (theanmolsharma) [inline packages/core/test/fixtures/transaction.ts:1]: Please double-check all the test vectors and make sure they are correct.

**Review by theanmolsharma — CHANGES_REQUESTED** [review]

File: packages/core/src/transaction.ts (hunk 4 above)
Comment (emjshrx) [reply packages/core/src/transaction.ts:143]: I have created a new utility function for this

File: packages/core/test/fixtures/transaction.ts [hunk 5]
```diff
@@ -54,6 +57,10 @@ export const validTransactions = [
                     script: 'a9147ccb85f0ab2d599bc17246c98babd5a20b1cdc7687',
                 },
             ],
+            inPubkeys: [
+                '0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798',
+                null,
```
Comment (emjshrx) [inline packages/core/test/fixtures/transaction.ts:62]: The second input seems to be a P2SH or something complex which I havent handled

File: packages/core/test/transaction.spec.ts [hunk 6]
```diff
@@ -48,4 +48,34 @@ describe('Transaction', () => {
...
+        'P2SH P2WSH P2PKH',
+        'P2SH P2WSH P2PK',
+        'Standard transaction (14:2)',
+        'Multisig',
+        'P2WSH Multisig',
+        'P2SH P2WSH Multisig',
+        'Coinbase transaction w/ witness',
+        'P2SH Multisig',
+        'Coinbase transaction',
+    ];
```
Comment (emjshrx) [inline packages/core/test/transaction.spec.ts:67]: These are the transaction types that arent handled

File: packages/core/src/transaction.ts [hunk 7]
```diff
@@ -126,4 +126,45 @@ export class Transaction {
...
+            if (isPubKey(scriptVector[1])) {
+                return scriptVector[1];
+            }
+        }
```
Comment (theanmolsharma) [inline packages/core/src/transaction.ts:147]: Don't use nested `if` when simplification is possible.

```suggestion
        if (scriptVector.length === 2 && isPubKey(scriptVector[1])) {
            return scriptVector[1];
        }
```

File: packages/core/src/transaction.ts [hunk 8]
```diff
@@ -126,4 +126,45 @@ export class Transaction {
...
+
+        // p2wpkh
+        if (scriptVector.length == 0 && input.witness.length === 2) {
+            if (isPubKey(input.witness[1])) {
```
Comment (theanmolsharma) [inline packages/core/src/transaction.ts:151]: same here, don't use nested `if`

File: packages/core/src/transaction.ts [hunk 9]
```diff
@@ -126,4 +126,45 @@ export class Transaction {
...
+                return input.witness[1];
+            }
+        }
+        // p2sh-p2wpkh
```
Comment (theanmolsharma) [inline packages/core/src/transaction.ts:155]: nit: add a line break between 2 `if` blocks

File: packages/core/src/transaction.ts [hunk 10]
```diff
@@ -126,4 +126,45 @@ export class Transaction {
...
+        if (
+            input.script.length === 23 &&
+            scriptVector.length == 1 &&
+            scriptVector[0][0] === 0x00 &&
+            scriptVector[0][1] === 0x14 &&
+            input.witness.length === 2 &&
+            input.witness[1].length === 33
+        ) {
+            if (isPubKey(input.witness[1])) {
```
Comment (theanmolsharma) [inline packages/core/src/transaction.ts:164]: same comment here regarding nested `if`

File: packages/core/src/utility.ts [hunk 11]
```diff
@@ -79,3 +79,10 @@ export const readVarInt = (buffer: Buffer, offset: number = 0): number => {
 export const encodingLength = (n: number) => {
     return n < 0xfd ? 1 : n <= 0xffff ? 3 : n <= 0xffffffff ? 5 : 9;
 };
+
+export const isPubKey = (testVector: Buffer | null): boolean => {
+    if (testVector && testVector.length == 33) {
```
Comment (theanmolsharma) [inline packages/core/src/utility.ts:84]: here we should also verify the public key using `secp256k1` to confirm if it's an actual public key and not some 33 byte random string

**Review by theanmolsharma — CHANGES_REQUESTED** [review]
Please make these changes and update tests for edge cases like "Invalid public key".
Also please sign your commits

**Comment by emjshrx** [comment]
> Please make these changes and update tests for edge cases like "Invalid public key". Also please sign your commits

I have added separate unit tests for `isPubKey` fn

File: packages/core/src/utility.ts [hunk 12]
```diff
@@ -79,3 +79,14 @@ export const readVarInt = (buffer: Buffer, offset: number = 0): number => {
 export const encodingLength = (n: number) => {
     return n < 0xfd ? 1 : n <= 0xffff ? 3 : n <= 0xffffffff ? 5 : 9;
 };
+
+export const isPubKey = (testVector: Buffer | null): boolean => {
```
Comment (theanmolsharma) [inline packages/core/src/utility.ts:83]: is there any reason why it should be allowed to accept `null` values?

File: packages/core/src/transaction.ts [hunk 13]
```diff
@@ -126,4 +126,45 @@ export class Transaction {
...
+        }
+
+        // p2sh-p2wpkh
+        if (
+            input.script.length === 23 &&
+            scriptVector.length == 1 &&
+            scriptVector[0][0] === 0x00 &&
+            scriptVector[0][1] === 0x14 &&
+            input.witness.length === 2 &&
+            input.witness[1].length === 33 &&
```
Comment (theanmolsharma) [inline packages/core/src/transaction.ts:163]: this is redundant with check inside `isPubKey()`

File: packages/core/test/fixtures/transaction.ts [hunk 14]
```diff
@@ -468,6 +487,9 @@ export const validTransactions = [
                     script: '76a914851a33a5ef0d4279bd5854949174e2c65b1d450088ac',
                 },
             ],
+            inPubkeys: [
+                '3045022100d269531f120f377ed2f94f42bef893ff2fe6544ac97fb477fa291bc6cfb7647e02200983f6a5bbd4ce6cf97f571995634805a7324cc5d8353ed954fa62477b0fcd0901',
```
Comment (theanmolsharma) [inline packages/core/test/fixtures/transaction.ts:491]: i don't think this is a public key. How are the tests passing?

File: packages/core/test/fixtures/transaction.ts [hunk 15]
```diff
@@ -293,6 +303,9 @@ export const validTransactions = [
                     script: '76a914851a33a5ef0d4279bd5854949174e2c65b1d450088ac',
                 },
             ],
+            inPubkeys: [
+                '21038de63cf582d058a399a176825c045672d5ff8ea25b64d28d4375dcdb14c02b2bac',
```
Comment (theanmolsharma) [inline packages/core/test/fixtures/transaction.ts:307]: this is also not a pub key

File: packages/core/test/fixtures/transaction.ts [hunk 16]
```diff
@@ -319,6 +332,9 @@ export const validTransactions = [
...
```
Comment (theanmolsharma) [inline packages/core/test/fixtures/transaction.ts:336]: same here

**Review by theanmolsharma — CHANGES_REQUESTED** [review]

File: packages/core/src/utility.ts (hunk 12 above)
Comment (emjshrx) [reply packages/core/src/utility.ts]: removed

File: packages/core/src/transaction.ts (hunk 13 above)
Comment (emjshrx) [reply packages/core/src/transaction.ts]: removed

File: packages/core/test/fixtures/transaction.ts (hunk 14 above)
Comment (emjshrx) [reply packages/core/test/fixtures/transaction.ts]: This was taken from a multisig address and this test has been skipped, removed the pubkey test data as its misleading

File: packages/core/test/fixtures/transaction.ts (hunk 15 above)
Comment (emjshrx) [reply packages/core/test/fixtures/transaction.ts]: This was taken from a P2SH address and this test has been skipped, removed the pubkey test data as its misleading

File: packages/core/test/fixtures/transaction.ts (hunk 16 above)
Comment (emjshrx) [reply packages/core/test/fixtures/transaction.ts]: This was taken from a P2SH address and this test has been skipped, removed the pubkey test data as its misleading

File: packages/core/test/fixtures/transaction.ts (hunk 14 above)
Comment (emjshrx) [reply packages/core/test/fixtures/transaction.ts]: I have added a test as well to make sure `isPubKey` returns false for these types of addresses

File: packages/core/src/utility.ts [hunk 17]
```diff
@@ -79,3 +79,14 @@ export const readVarInt = (buffer: Buffer, offset: number = 0): number => {
 export const encodingLength = (n: number) => {
     return n < 0xfd ? 1 : n <= 0xffff ? 3 : n <= 0xffffffff ? 5 : 9;
 };
+
+export const isPubKey = (testVector: Buffer): boolean => {
+    if (
+        testVector &&
```
Comment (theanmolsharma) [inline packages/core/src/utility.ts:85]: i don't think we need to do this

File: packages/core/src/utility.ts [hunk 18]
```diff
@@ -79,3 +79,14 @@ export const readVarInt = (buffer: Buffer, offset: number = 0): number => {
...
+        (testVector.length == 33 || testVector.length == 65) &&
+        secp256k1.publicKeyVerify(testVector)
+    ) {
+        return true;
+    }
+    return false;
```
Comment (theanmolsharma) [inline packages/core/src/utility.ts:91]:
```suggestion

        return (testVector.length == 33 || testVector.length == 65) &&
        secp256k1.publicKeyVerify(testVector);
```

File: packages/core/test/transaction.spec.ts [hunk 19]
```diff
@@ -48,4 +48,34 @@ describe('Transaction', () => {
...
+
+    it.each(
+        validTransactions.filter(
```
Comment (theanmolsharma) [inline packages/core/test/transaction.spec.ts:70]: why do we need to do this filtering? shouldn't `getPublicKeyFromInput()` return `null` for all cases that are being filtered out?

**Review by theanmolsharma — COMMENTED** [review]
I think we are good to merge after this iteration. Thanks for being patient.

File: packages/core/test/transaction.spec.ts (hunk 19 above)
Comment (emjshrx) [reply packages/core/test/transaction.spec.ts:70]: That would mean I have to add `null` to fixtures as the `inPubkeys` which seemed wrong to me as it could be a P2SH tx or something which has a PK. It's just that currently we only support a few txs, and the null cases should not be tested against IMO. What do you think?

Comment (theanmolsharma) [reply packages/core/test/transaction.spec.ts:70]: fair point

**Review by theanmolsharma — APPROVED** [review]

**Comment by theanmolsharma** [comment]
@emjshrx please squash 29ff2f367171125d0fa19849af613b000ed504a8 into 7f6b201d6a34a8df100a526b3e6336b2b68827b0

**Comment by emjshrx** [comment]
@theanmolsharma apologies for the 2 month delay, this is now ready to be merged :). Also add a small commit to fix the CI
